flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/rank/UpdatableTopNFunction.java

```diff
@@ -541,8 +541,10 @@ private void processElementWithoutRowNumber(RowData inputRow, Collector<RowData>
         }
     }
 
-    private void flushBufferToState(Tuple2<TopNBuffer, Map<RowData, RankRow>> bufferEntry)
+    private void flushBufferToState(
```

Comment:
I'd suggest providing a more complete cache also tracking those keys that are already removed from the cache map but still flushing to the state, to maintain cache coherence for accessing during flushing

Reply:
As it stands, there is no issue with the append top N. I have placed it under the Jira ticket for introducing the updatable top N with the async state API https://issues.apache.org/jira/browse/FLINK-36936